packages/storefront-sdk/src/client/index.ts

```diff
@@ -1,19 +1,125 @@
 import { createClient } from '@urql/core';
+import { errorMessages, dataFetchingMethods } from '../utils/index.js';
 import type { Client as UrqlClient } from '@urql/core';
 import type { StorefrontClientParams } from '../index.js';
+import type {
+	AfterCallback,
+	AfterCallbackWithId,
+	AfterSubscriptions,
+	DataFetchingMethodName,
+	MethodData
+} from '../types/after.js';
+import type { Content } from '../types/storefront.js';
 
 export class StorefrontClient {
 	readonly #graphqlClient: UrqlClient;
+	readonly #afterSubscriptions: AfterSubscriptions<DataFetchingMethodName>;
 
 	constructor(params: StorefrontClientParams) {
 		this.#graphqlClient = createClient({
 			url: params.storefrontEndpoint,
 			fetch: params.fetchClient ?? globalThis.fetch
 		});
+		this.#afterSubscriptions = {};
+	}
+
+	/**
+	 * TODO: Remove this and refactor `.after` tests once the `.getConfig` method is available.
+	 * Once we have access to `.getConfig`, we can just:
+	 * ```
+	 * const { afterSubscriptions } = client.getConfig();
+	 * ```
+	 */
+	get afterSubscriptions() {
+		return this.#afterSubscriptions;
+	}
+
+	/**
+	 * Register a callback function that gets applied to the return value of a data-fetching method. When a callback is registered, it is applied every time the data-fetching method runs.
+	 * @param method The method to apply the `callback` to. Possible values: `'content'`, `'navigation'`, `'products'`, `'productCollections'`, `'productCollectionEntries'`, `'query'`, and `'spaceProperties'`.
+	 * @param callback The callback that gets applied to the data returned by the `method` of interest.
+	 * @param callbackId Optional ID. If not provided, an ID will be assigned. A `callback` will be overwritten if a new `callback` is registered to the same `method` with the same `callbackId`.
+	 */
+	after<MethodName extends DataFetchingMethodName>(
+		method: MethodName,
+		callback: AfterCallback<MethodData[MethodName]> | null,
+		callbackId?: string
+	) {
+		if (!dataFetchingMethods.includes(method)) {
+			throw new Error(errorMessages.afterMethodInvalid(method));
+		}
+
+		if (
+			typeof callback === 'undefined' ||
+			(typeof callback !== 'function' && callback !== null)
+		) {
+			throw new Error(errorMessages.afterMethodCallbackInvalid(callback));
+		}
+
+		if (typeof callbackId !== 'undefined' && typeof callbackId !== 'string') {
+			throw new Error(errorMessages.afterMethodCallbackIdInvalid(callbackId));
+		}
+
+		let methodSubscriptions = this.#afterSubscriptions[method];
+
+		if (typeof methodSubscriptions === 'undefined') {
+			methodSubscriptions = this.#afterSubscriptions[method] = {};
+		}
+
+		const id =
+			callbackId ?? `${method}::${Object.values(methodSubscriptions).length}`;
+
+		if (typeof callback === 'function') {
+			(methodSubscriptions as unknown as AfterCallbackWithId<MethodName>)[id] =
```

Comment:
This type assertion bothers me a little. But without it, TypeScript seems unable to understand what the type should be.

Reply:
It seems like we should be able to do away with this assertion at some stage. This is the error TS gives, which makes me think that there's something little we can fix since it's just acting like the two types don't match despite having virtually identical signatures: 

```sh
 Type 'AfterCallback<MethodData[MethodName]>' is not assignable to type 'AfterCallback<Content[] | ContentEdge[] | Product[] | ProductCollectionEdge[] | ProductEdge[] | NavigationGroup[] | ProductCollection[] | SpaceProperties | StorefrontResponse<...>>'.
  Types of parameters 'responseObj' and 'responseObj' are incompatible.
    Type 'Content[] | ContentEdge[] | Product[] | ProductCollectionEdge[] | ProductEdge[] | NavigationGroup[] | ProductCollection[] | SpaceProperties | StorefrontResponse<...>' is not assignable to type 'MethodData[MethodName]'.
```

Reply:
At least part of the problem seems to be that TS doesn't know that `this.#afterSubscriptions[method]` cannot be `undefined` on this line, despite the assignment above:

```ts
if (typeof methodSubscriptions === 'undefined') {
	methodSubscriptions = this.#afterSubscriptions[method] = {};
}
```

If you remove the assertion but keep parentheses, we get:

```
Object is possibly 'undefined'. ts(2532)
```

at the `(methodSubscriptions)[id]` assignment.

![object is possibly undefined](https://user-images.githubusercontent.com/5732000/210897700-f380b81a-72cb-41a5-a08c-25a1ce4aa62b.png)

It might have something to do with https://github.com/microsoft/TypeScript/issues/27706.

Reply:
It seems like there's something subtly off about the assignment type for sure. With some non-null assertions and string type casts to make sure it recognize the id type correctly, I still get this: 

```sh
Type 'AfterCallback<MethodData[MethodName]>' is not assignable to type 'NonNullable<Partial<Record<"query" | "navigation" | "content" | "products" | "productCollections" | "productCollectionEntries" | "spaceProperties", AfterCallbackWithId<"query" | "navigation" | "content" | "products" | "productCollections" | "productCollectionEntries" | "spaceProperties">>>[MethodName]>[MethodName]'
```